proxy/gateway/java/nanomsg-binding/java_nanomsg.c

```
@@ -0,0 +1,197 @@
+}
+
+JNIEXPORT jint JNICALL Java_com_microsoft_azure_gateway_remote_NanomsgLibrary_nn_1close
+(JNIEnv *env, jobject obj, jint socket) {
+    return nn_close(socket);
+}
+
+JNIEXPORT jint JNICALL Java_com_microsoft_azure_gateway_remote_NanomsgLibrary_nn_1bind
+(JNIEnv *env, jobject obj, jint socket, jstring address) {
+    const char *addr = (*env)->GetStringUTFChars(env, address, NULL);
+    
+    if (addr == NULL)
+    {
+        addr = 0;
+    }
+    
+    return nn_bind(socket, addr);
+}
+
+JNIEXPORT jint JNICALL Java_com_microsoft_azure_gateway_remote_NanomsgLibrary_nn_1shutdown
+(JNIEnv *env, jobject obj, jint socket, jint endpoint) {
+    return nn_shutdown(socket, endpoint);
+}
+
+JNIEXPORT jint JNICALL Java_com_microsoft_azure_gateway_remote_NanomsgLibrary_nn_1send
+(JNIEnv *env, jobject obj, jint socket, jbyteArray buffer, jint flags) {
+    jint result = -1;
+    jsize length = (*env)->GetArrayLength(env, buffer);
+    
+    jbyte* cbuffer = (*env)->GetByteArrayElements(env, buffer, 0);
+        
+    if (cbuffer == NULL)
+    {
+        result = -1;
+    }
+    else {
+        result = nn_send(socket, cbuffer, length, flags);
+        (*env)->ReleaseByteArrayElements(env, buffer, cbuffer, JNI_ABORT);
+    }
+
+    return result;
+}
+
+
+JNIEXPORT jbyteArray JNICALL Java_com_microsoft_azure_gateway_remote_NanomsgLibrary_nn_1recv
+(JNIEnv *env, jobject obj, jint socket, jint flags)
+{
+    void *buf = NULL;
+    int nbytes = nn_recv(socket, &buf, NN_MSG, flags);
+    jbyteArray result = 0;
+    if (nbytes < 0) {
+        result = 0;
+    }
+    else {
+        result = (*env)->NewByteArray(env, nbytes);
+        
+        if (result == NULL) {
+            result = 0;
+        }
+        else {
+            (*env)->SetByteArrayRegion(env, result, 0, nbytes, (const jbyte*)buf);
+            jthrowable exception = (*env)->ExceptionOccurred(env);
+            if (exception) {
+                result = 0;
+                (*env)->ExceptionClear(env);
```

I am not sure, but are we sort of swallowing the exception here? Should this not get propagated back to the caller?